ESPS/general/man/man3/genzfunc.3
.\" Copyright (c) 1987, 1990 Entropic Speech, Inc.
.\" Copyright (c) 1991 Entropic Research Laboratory, Inc.; All rights reserved
.\" @(#)genzfunc.3	1.8 29 Apr 1997 ESI/ERL
.ds ]W (c) 1991 Entropic Research Laboratory, Inc.
.TH GENZFUNC 3\-ESPSu 29 Apr 1997
.ds ]W "\fI\s+4\ze\h'0.05'e\s-4\v'-0.4m'\fP\(*p\v'0.4m'\ Entropic Research Laboratory, Inc.
.SH NAME
.nf
add_genzfunc \- create generic header items to store a filter (zfunc)
get_genzfunc \- obtain a filter (zfunc) from generic header items 
.SH SYNOPSIS
.nf
.ft B
#include <esps/esps.h>

void
add_genzfunc(name, filter, hd)
char *name;		
struct zfunc *filter;
struct header *hd;	

struct zfunc *
get_genzfunc(name, hd)
char *name;		
struct header *hd;	
.ft

.SH DESCRIPTION
.fi
.I add_genzfunc
creates three generic header items to store the information in the 
zfunc 
.I filter,
and it stores the information in those header items.  
.I get_genzfunc
obtains information from three generic header items in \fIhd\fP,
turns this into a zfunc, and returns a pointer to the zfunc.  
The names of the generic header items are based on the "basename" 
.I name.  
The number of numerator coefficients
and denominator coefficients are stored 
as two values in "\fIname_\fPsiz" as shorts.
These coefficients are normally the transfer function coefficients of
a digital filter.
The numerator coefficients are stored in "\fIname\fP_zeros" and the 
denominator coefficients
are stored in "\fIname\fP_poles" as floats.  
.PP
.I get_genzfunc
returns NULL if "\fIname_\fPsiz" is not a defined SHORT generic header item
or its size is not 2.  NULL is also returned if "\fIname\fP_zeros" 
or "\fIname\fP_poles" is not a defined FLOAT generic or has a size 
different from that specified in "\fIname_\fPsiz".  
.SH EXAMPLE
.PP
.nf
struct header *ana_ih; /*ANA input header*/
struct header *fea_oh; /*FEA output header*/
add_genzfunc("pre_emp", ana_ih->hd.ana->pre_emp, fea_oh);
.PP
.nf
struct header *fea_ih; /*FEA input header*/
struct header *ana_oh; /*ANA output header*/
ana_oh->hd.ana->pre_emp = get_genzfunc("pre_emp", fea_ih);
.fi
.SH BUGS
None known.
.SH "SEE ALSO"
.nf
\fIgenhd_codes\fR(3-ESPSu), \fIgenhd_list\fR(3\-ESPSu), \fIgenhd_type\fR(3\-ESPSu)
.fi
.SH AUTHOR
John Shore  


ESPS/general/src/lib/genzfunc.c
/*
 * This material contains proprietary software of Entropic Speech, Inc.
 * Any reproduction, distribution, or publication without the prior
 * written permission of Entropic Speech, Inc. is strictly prohibited.
 * Any public distribution of copies of this work authorized in writing by
 * Entropic Speech, Inc. must bear the notice
 *
 *    "Copyright (c) 1987, 1990 Entropic Speech, Inc.; All rights reserved"
*/
#ifndef lint
static char *sccs_id = "@(#)genzfunc.c	1.7	10/19/93 ESI";
#endif

#include <stdio.h>
#include <esps/esps.h>

#ifndef DEC_ALPHA
char *malloc(), *strcat(), *strcpy();
void free();
#endif
void f_copy();


void
add_genzfunc(name, filter, hd)
char *name;			/* base name of filter */
struct zfunc *filter;			/* filter to add */
struct header *hd;		/* SPS file header */
{
/* this function takes a zfunc and a basename (name) and it creates
 * generic header items to store the filter information; the number 
 * of zeros and poles are stored in name_siz, the zeros in name_zeros,
 * and the poles in name_poles; the inverse function is get_genhd_z()
 */
    short *pre_siz;
    float *pre_zeros, *pre_poles;
    char *basename;		/* base name of filter */

/* check arguments
*/
    spsassert(name != NULL,"add_genzfunc: name NULL");
    spsassert(hd != NULL,"add_genzfunc: hd NULL");
    spsassert(filter != NULL,"add_genzfunc: filter NULL");

    basename = malloc((unsigned)strlen(name) + 7);
    spsassert(basename != NULL, "add_genzfunc: memory allocation failed");

    (void) strcpy(basename, name);
    pre_siz = add_genhd_s(strcat(basename, "_siz"), (short *)NULL, 2, hd);
    pre_siz[0] = filter->nsiz;
    pre_siz[1] = filter->dsiz;

    if(pre_siz[0]) {
    (void) strcpy(basename, name);
    pre_zeros = add_genhd_f(strcat(basename, "_zeros"), 
			    filter->zeros, (int)pre_siz[0], hd);
    }

    if(pre_siz[1]) {
    (void) strcpy(basename, name);
    pre_poles = add_genhd_f(strcat(basename, "_poles"), 
			    filter->poles, (int)pre_siz[1], hd);

    }
    free(basename);
}


struct zfunc *
get_genzfunc(name, hd)
char *name;			/* base name of zfunc */
struct header *hd;		/* pointer to SPS header */
{
/* this function accepts the basename (name) of a filter, reads 
 * SPS header items for the number of poles and zeros and also 
 * for the poles and zeros themselves; this information is converted 
 * to a zfunc and returned.
 */
    short *pre_siz;
    float *pre_zero, *pre_pol;
    int size, gotzfunc = 0;
    char *basename_siz, *basename_poles, *basename_zeros;		

/* check arguments
*/
    spsassert(name != NULL,"get_genzfunc: name NULL");
    spsassert(hd != NULL,"get_genzfunc: hd NULL");

    basename_siz = malloc((unsigned)strlen(name) + 5);
    spsassert(basename_siz != NULL, "get_genzfunc: memory allocation failed");

    (void) strcpy(basename_siz, name);
    (void) strcat(basename_siz, "_siz");

    if (genhd_type(basename_siz, &size, hd) == SHORT && size == 2) {
	gotzfunc = 1;
	basename_poles = malloc((unsigned)strlen(name) + 7);
	basename_zeros = malloc((unsigned)strlen(name) + 7);

	spsassert(basename_poles != NULL, 
		"get_genzfunc: memory allocation failed");
	spsassert(basename_zeros != NULL, 
		"get_genzfunc: memory allocation failed");

	(void) strcpy(basename_poles, name);
	(void) strcat(basename_poles, "_poles");
	(void) strcpy(basename_zeros, name);
	(void) strcat(basename_zeros, "_zeros");
    	pre_siz = (short *)get_genhd(basename_siz, hd);

    	if (pre_siz[0] > 0 
	      && genhd_type(basename_zeros, &size, hd) == FLOAT
	      && size == pre_siz[0])
 	    pre_zero = (float *)get_genhd(basename_zeros, hd);
	else
	    gotzfunc = 0;

	free(basename_zeros);

    	if (pre_siz[1] > 0 
	      && genhd_type(basename_poles, &size, hd) == FLOAT
	      && size == pre_siz[1])
	    pre_pol = (float *)get_genhd(basename_poles, hd);
	else
	    gotzfunc = 0;

	free(basename_poles);
    }

    free(basename_siz);

    if (gotzfunc) 
    	return new_zfunc(pre_siz[0], pre_siz[1], pre_zero, pre_pol);
    else
	return NULL;
}


